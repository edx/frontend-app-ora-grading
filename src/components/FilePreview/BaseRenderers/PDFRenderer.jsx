import React from 'react';
import PropTypes from 'prop-types';

import { pdfjs, Document, Page } from 'react-pdf';
import {
  Icon, Form, ActionRow, IconButton,
} from '@edx/paragon';
import { ChevronLeft, ChevronRight } from '@edx/paragon/icons';
import pdfjsWorker from 'react-pdf/dist/esm/pdf.worker.entry';

import 'react-pdf/dist/esm/Page/AnnotationLayer.css';

pdfjs.GlobalWorkerOptions.workerSrc = pdfjsWorker;

/**
 * <PDFRenderer />
 */
export class PDFRenderer extends React.Component {
  static INITIAL_STATE = {
    pageNumber: 1,
    numPages: 1,
    relativeHeight: 0,
  };

  constructor(props) {
    super(props);

    this.state = { ...PDFRenderer.INITIAL_STATE };

    this.wrapperRef = React.createRef();
    this.onDocumentLoadSuccess = this.onDocumentLoadSuccess.bind(this);
    this.onDocumentLoadError = this.onDocumentLoadError.bind(this);
    this.onLoadPageSuccess = this.onLoadPageSuccess.bind(this);
    this.onPrevPageButtonClick = this.onPrevPageButtonClick.bind(this);
    this.onNextPageButtonClick = this.onNextPageButtonClick.bind(this);
    this.onInputPageChange = this.onInputPageChange.bind(this);
  }

  onDocumentLoadSuccess = ({ numPages }) => {
    this.props.onSuccess();
    this.setState({ numPages });
  };

  onLoadPageSuccess = (page) => {
    const pageWidth = page.view[2];
    const pageHeight = page.view[3];
    const wrapperHeight = this.wrapperRef.current.getBoundingClientRect().width;
    const relativeHeight = (wrapperHeight * pageHeight) / pageWidth;
    if (relativeHeight !== this.state.relativeHeight) {
      this.setState({ relativeHeight });
    }
  };

  onDocumentLoadError = (error) => {
    let status;
    switch (error.name) {
      case 'MissingPDFException':
        status = 404;
        break;
      default:
        status = 500;
        break;
    }
    this.props.onError(status);
  };

  onInputPageChange = ({ target: { value } }) => {
    this.setPageNumber(parseInt(value, 10));
  }

  onPrevPageButtonClick = () => {
    this.setPageNumber(this.state.pageNumber - 1);
  }

  onNextPageButtonClick = () => {
    this.setPageNumber(this.state.pageNumber + 1);
  }

  setPageNumber(pageNumber) {
    if (pageNumber > 0 && pageNumber <= this.state.numPages) {
      this.setState({ pageNumber });
    }
  }

  get hasNext() {
    return this.state.pageNumber < this.state.numPages;
  }

  get hasPrev() {
    return this.state.pageNumber > 1;
  }

  render() {
    return (
      <div ref={this.wrapperRef} className="pdf-renderer">
        <Document
          file={this.props.url}
          onLoadSuccess={this.onDocumentLoadSuccess}
          onLoadError={this.onDocumentLoadError}
        >
          {/* <Outline /> */}
          <div
            className="page-wrapper"
            style={{
              height: this.state.relativeHeight,
            }}
          >
            <Page
              pageNumber={this.state.pageNumber}
              onLoadSuccess={this.onLoadPageSuccess}
            />
          </div>
        </Document>
        <ActionRow className="d-flex justify-content-center m-0">
          <IconButton
            size="inline"
            alt="previous pdf page"
            iconAs={Icon}
            src={ChevronLeft}
            disabled={!this.hasPrev}
            onClick={this.onPrevPageButtonClick}
          />
          <Form.Group className="d-flex align-items-center m-0">
            <Form.Label isInline>Page </Form.Label>
            <Form.Control
              type="number"
              min={0}
              max={this.state.numPages}
              value={this.state.pageNumber}
              onChange={this.onInputPageChange}
            />
            <Form.Label isInline> of {this.state.numPages}</Form.Label>
          </Form.Group>
          <IconButton
            size="inline"
            alt="next pdf page"
            iconAs={Icon}
            src={ChevronRight}
            disabled={!this.hasNext}
            onClick={this.onNextPageButtonClick}
          />
        </ActionRow>
      </div>
    );
  }
}

PDFRenderer.defaultProps = {};

PDFRenderer.propTypes = {
  url: PropTypes.string.isRequired,
  onError: PropTypes.func.isRequired,
  onSuccess: PropTypes.func.isRequired,
};

export default PDFRenderer;
