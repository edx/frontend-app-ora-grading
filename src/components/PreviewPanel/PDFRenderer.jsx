import React from 'react';
import PropTypes from 'prop-types';

import { Document, Page } from 'react-pdf/dist/esm/entry.webpack'; // avoid import worker directly

import {
  Icon, Form, ActionRow, IconButton,
} from '@edx/paragon';
import { ChevronLeft, ChevronRight } from '@edx/paragon/icons';

/**
 * <PDFRenderer />
 */
export class PDFRenderer extends React.Component {
  static supportedTypes = ['pdf'];

  constructor(props) {
    super(props);

    this.state = {
      pageNumber: 1,
      numPages: 1,
      relativeHeight: 0,
    };

    this.wrapperRef = React.createRef();
    this.onDocumentLoadSuccess = this.onDocumentLoadSuccess.bind(this);
    this.onDocumentLoadError = this.onDocumentLoadError.bind(this);
    this.onLoadPageSuccess = this.onLoadPageSuccess.bind(this);
  }

  onDocumentLoadSuccess = ({ numPages }) => {
    this.setState({ numPages });
  };

  onLoadPageSuccess = (page) => {
    // 686 * 1078 / 1862
    const pageWidth = page.view[2];
    const pageHeight = page.view[3];
    const wrapperHeight = this.wrapperRef.current.getBoundingClientRect().width;
    const relativeHeight = (wrapperHeight * pageHeight) / pageWidth;
    if (relativeHeight !== this.state.relativeHeight) {
      this.setState({ relativeHeight });
    }
  };

  onDocumentLoadError = (error) => {
    // eslint-disable-next-line no-console
    console.error(error);
  };

  setPageNumber(pageNumber) {
    if (pageNumber > 0 && pageNumber <= this.state.numPages) {
      this.setState({ pageNumber });
    }
  }

  get hasNext() {
    return this.state.pageNumber < this.state.numPages;
  }

  get hasPrev() {
    return this.state.pageNumber > 1;
  }

  render() {
    return (
      <div ref={this.wrapperRef} className="pdf-renderer">
        <Document
          file={this.props.url}
          onLoadSuccess={this.onDocumentLoadSuccess}
          onLoadError={this.onDocumentLoadError}
        >
          {/* <Outline /> */}
          <div
            className="page-wrapper"
            style={{
              height: this.state.relativeHeight,
            }}
          >
            <Page
              pageNumber={this.state.pageNumber}
              onLoadSuccess={this.onLoadPageSuccess}
            />
          </div>
        </Document>
        <ActionRow className="d-flex justify-content-center m-0">
          <IconButton
            size="inline"
            alt="previous pdf page"
            iconAs={Icon}
            src={ChevronLeft}
            disabled={!this.hasPrev}
            onClick={() => this.setPageNumber(this.state.pageNumber - 1)}
          />
          <Form.Group className="d-flex align-items-center m-0">
            <Form.Label isInline>Page </Form.Label>
            <Form.Control
              type="number"
              min={0}
              max={this.state.numPages}
              value={this.state.pageNumber}
              onChange={(e) => this.setPageNumber(parseInt(e.target.value, 10))}
            />
            <Form.Label isInline> of {this.state.numPages}</Form.Label>
          </Form.Group>
          <IconButton
            size="inline"
            alt="next pdf page"
            iconAs={Icon}
            src={ChevronRight}
            disabled={!this.hasNext}
            onClick={() => this.setPageNumber(this.state.pageNumber + 1)}
          />
        </ActionRow>
      </div>
    );
  }
}

PDFRenderer.defaultProps = {};

PDFRenderer.propTypes = {
  url: PropTypes.string.isRequired,
};

export default PDFRenderer;
