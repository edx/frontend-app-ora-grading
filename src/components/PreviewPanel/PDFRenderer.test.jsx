import React from 'react';
import { shallow } from 'enzyme';

import PDFRenderer from './PDFRenderer';

describe('PDF Renderer Component', () => {
  const props = {
    url: 'some_url.pdf',
  };

  let el;
  beforeEach(() => {
    el = shallow(<PDFRenderer {...props} />);
    el.instance().onDocumentLoadSuccess = jest
      .fn()
      .mockName('onDocumentLoadSuccess');
    el.instance().onDocumentLoadError = jest
      .fn()
      .mockName('onDocumentLoadError');
    el.instance().onLoadPageSuccess = jest.fn().mockName('onLoadPageSuccess');
  });
  test('snapshot', () => {
    el.instance().onPrevPageButtonClick = jest
      .fn()
      .mockName('onPrevPageButtonClick');
    el.instance().onNextPageButtonClick = jest
      .fn()
      .mockName('onNextPageButtonClick');
    el.instance().onInputPageChange = jest.fn().mockName('onInputPageChange');
    expect(el.instance().render()).toMatchSnapshot();
  });

  describe('Component', () => {
    const supportedTypes = ['pdf'];
    test('static supported types is expected', () => {
      expect(PDFRenderer.supportedTypes).toEqual(supportedTypes);
    });

    describe('state', () => {
      const state = { numPages: 10 };
      describe('previous page button', () => {
        test('enable', () => {
          el.setState({ pageNumber: state.numPages, ...state });
          const { hasPrev } = el.instance();
          const PrevButtonEl = el.find('IconButton').at(0);
          expect(hasPrev).toEqual(true);
          expect(PrevButtonEl.prop('disabled')).toEqual(!hasPrev);
        });
        test('disabled', () => {
          el.setState({ pageNumber: 1, ...state });
          const { hasPrev } = el.instance();
          const PrevButtonEl = el.find('IconButton').at(0);
          expect(hasPrev).toEqual(false);
          expect(PrevButtonEl.prop('disabled')).toEqual(!hasPrev);
        });

        test('onPrevPageButtonClick', () => {
          el.setState({ pageNumber: state.numPages, ...state });
          const PrevButtonEl = el.find('IconButton').at(0);
          PrevButtonEl.simulate('click');
          expect(el.state('pageNumber')).toEqual(state.numPages - 1);
        });
      });

      describe('next page button', () => {
        test('enable', () => {
          el.setState({ pageNumber: 1, ...state });
          const { hasNext } = el.instance();
          expect(hasNext).toEqual(true);
          const NextButtonEl = el.find('IconButton').at(1);
          expect(NextButtonEl.prop('disabled')).toEqual(!hasNext);
        });
        test('disabled', () => {
          el.setState({ pageNumber: state.numPages, ...state });
          const { hasNext } = el.instance();
          expect(hasNext).toEqual(false);
          const NextButtonEl = el.find('IconButton').at(1);
          expect(NextButtonEl.prop('disabled')).toEqual(!hasNext);
        });
        test('onNextPageButtonClick', () => {
          const pageNumber = 1;
          el.setState({ pageNumber, ...state });
          const NextButtonEl = el.find('IconButton').at(1);
          NextButtonEl.simulate('click');
          expect(el.state('pageNumber')).toEqual(pageNumber + 1);
        });
      });

      describe('input page change', () => {
        test('onInputPageChange', () => {
          el.setState({ pageNumber: state.numPages, ...state });
          const event = {
            target: {
              value: 5,
            },
          };
          el.find('[type="number"]').simulate('change', event);
          el.instance().onInputPageChange(event);
          expect(el.state('pageNumber')).toEqual(event.target.value);
          expect(el.state('pageNumber')).not.toEqual(state.numPages);
        });
      });
    });
  });
});
