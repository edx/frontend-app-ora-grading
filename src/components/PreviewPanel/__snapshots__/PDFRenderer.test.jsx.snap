// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PDF Renderer Component snapshot 1`] = `
<div
  className="pdf-renderer"
>
  <Document
    error="Failed to load PDF file."
    file="some_url.pdf"
    loading="Loading PDFâ€¦"
    noData="No PDF file specified."
    onLoadError={[MockFunction onDocumentLoadError]}
    onLoadSuccess={[MockFunction onDocumentLoadSuccess]}
    onPassword={[Function]}
  >
    <div
      className="page-wrapper"
      style={
        Object {
          "height": 0,
        }
      }
    >
      <ForwardRef(Page)
        onLoadSuccess={[MockFunction onLoadPageSuccess]}
        pageNumber={1}
      />
    </div>
  </Document>
  <ActionRow
    className="d-flex justify-content-center m-0"
  >
    <IconButton
      alt="previous pdf page"
      disabled={true}
      iconAs="Icon"
      onClick={[MockFunction onPrevPageButtonClick]}
      size="inline"
      src={[MockFunction icons.ChevronLeft]}
    />
    <Form.Group
      className="d-flex align-items-center m-0"
    >
      <Form.Label
        isInline={true}
      >
        Page 
      </Form.Label>
      <Form.Control
        max={1}
        min={0}
        onChange={[MockFunction onInputPageChange]}
        type="number"
        value={1}
      />
      <Form.Label
        isInline={true}
      >
         of 
        1
      </Form.Label>
    </Form.Group>
    <IconButton
      alt="next pdf page"
      disabled={true}
      iconAs="Icon"
      onClick={[MockFunction onNextPageButtonClick]}
      size="inline"
      src={[MockFunction icons.ChevronRight]}
    />
  </ActionRow>
</div>
`;
